Add linalg.lu_solve

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73878
* #73877
* #74046
* #77637
* #77636
* #77635
* __->__ #77634

Relanding https://github.com/pytorch/pytorch/pull/72935

Differential Revision: [D36793144](https://our.internmc.facebook.com/intern/diff/D36793144)

lezcano committed:
Add linalg.lu_solve

This PR adds `linalg.lu_solve`. While doing so, I found a bug in MAGMA
when calling the batched MAGMA backend with trans=True. We work around
that by solving the system solving two triangular systems.

We also update the heuristics for this function, as they were fairly
updated. We found that cuSolver is king, so luckily we do not need to
rely on the buggy backend from magma for this function.

We added tests testing this function left and right. We also added tests
for the different backends. We also activated the tests for AMD, as
those should work as well.

Fixes https://github.com/pytorch/pytorch/issues/61657

[ghstack-poisoned]

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

[ghstack-poisoned]

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

[ghstack-poisoned]

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

[ghstack-poisoned]

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

[ghstack-poisoned]

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

[ghstack-poisoned]

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

[ghstack-poisoned]

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

[ghstack-poisoned]

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

[ghstack-poisoned]

ngimel commented:
Do we need to run this on ci/trunk? Why was it reverted previously?

Lezcano commented:
It broke an internal build. See https://github.com/pytorch/pytorch/pull/72935#issuecomment-1121459787 https://github.com/pytorch/pytorch/pull/72935#issuecomment-1121470502 https://github.com/pytorch/pytorch/pull/72935#issuecomment-1121473646 https://github.com/pytorch/pytorch/pull/72935#issuecomment-1128109367

ngimel commented:
Ok, so what next? We can't just reland it because it'll break internal build again?

Lezcano commented:
According to https://github.com/pytorch/pytorch/pull/72935#issuecomment-1128109367, I'd assume that @malfet already corrected the internal error?

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36793144).

ngimel commented:
I'm still getting internal build failures on this one 
```
Summary: 
ld.lld: error: undefined symbol: at::native::DispatchStubImpl::get_call_ptr(c10::DeviceType, void*)
stderr: ld.lld: error: undefined symbol: at::native::DispatchStubImpl::get_call_ptr(c10::DeviceType, void*)
``` 
so we cannot land it.

malfet commented:
> @malfet any updates on this one?

Sorry, completely slipped my mind. Do you mind rebasing it against latest viable/strict and let me try the import again...

lezcano committed:
Update on "Add linalg.lu_solve"


Relanding https://github.com/pytorch/pytorch/pull/72935

Differential Revision: [D36793144](https://our.internmc.facebook.com/intern/diff/D36793144)

[ghstack-poisoned]

Lezcano commented:
@malfet rebased.
I needed to add an import that was missing to a file that was unrelated to this PR (idk why but it was not building locally...)

malfet commented:
Ok, I finally understand what is going on(torch_cuda in ideal world should be oblivious of supported CPU architectures, which is enforced internally but not in OSS build, so have an idea how to fix it internally):
https://github.com/pytorch/pytorch/blob/32461ed319866f06c95c1bedd61a6acbceca9529/aten/src/ATen/native/DispatchStub.h#L75-L80

Lezcano commented:
@pytorchbot merge

kit1980 commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:
```diff
@@ -3012,82 +3018,173 @@ static void apply_lu_solve_batched_magma(const Tensor& b, const Tensor& lu, cons
 
-}  // anonymous namespace
+static void lu_solve_kernel(const Tensor& LU, const Tensor& pivots, const Tensor& B, TransposeType trans) {
+  // Trivial case. Remove it once `torch.solve` is removed, as linalg.solve already shortcuts this case
```
@lezcano is it time to remove this now?

lezcano replied:
It's rather minor, but sure, happy to approve a PR that removes this provided that CI passes
